File: packages/mobile/android/app/src/main/java/id/homebase/photos/SyncTrigger.java
package id.homebase.photos;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactMethod;

import id.homebase.lib.core.DotYouClient;
import id.homebase.lib.core.file.types.BadRequestUploadResult;
import id.homebase.lib.core.file.types.SuccessfullUploadResult;
import id.homebase.lib.core.file.types.UploadResult;
import id.homebase.photos.mediasync.ImageProvider;
import id.homebase.photos.mediasync.MediaSync;
import id.homebase.photos.mediasync.VideoProvider;

public class SyncTrigger extends ReactContextBaseJavaModule {

    public SyncTrigger(ReactApplicationContext context) {
        super(context);
    }


    @ReactMethod
    public void runSync() {
        MediaSync mediaSync = new MediaSync(getReactApplicationContext());
        mediaSync.syncMedia();
    }

    @ReactMethod
    public void runSingleSync(String filePath, double timestampInMillis, String mimeType, String identifier, double width, double height, Promise promise) {
        DotYouClient dotYouClient = DotYouClient.getDotYouClient(this.getReactApplicationContext());
        try {
            UploadResult result;
            if (mimeType.startsWith("video/")) {
                VideoProvider videoProvider = new VideoProvider(this.getReactApplicationContext());
                result = videoProvider.uploadMedia(dotYouClient, filePath, (long) timestampInMillis, mimeType, identifier, String.valueOf(width), String.valueOf(height), true);
            } else {
                result = ImageProvider.uploadMedia(dotYouClient, filePath, (long) timestampInMillis, mimeType, identifier, String.valueOf(width), String.valueOf(height), false);
            }

            if (result instanceof SuccessfullUploadResult) {
                promise.resolve("Upload success");
            } else if(result instanceof BadRequestUploadResult) {
                promise.reject("BAD_REQUEST", "Upload failed", new Exception("Upload failed"));
            } else {
                promise.reject("UNKNOWN_ERROR", "Upload failed", new Exception("Upload failed"));
            }
        } catch (Exception e) {
            promise.reject("SYNC_ERROR", "Upload failed", e);
        }
    }

    @NonNull
    @Override
    public String getName() {
        return "SyncTrigger";
    }
}

File: packages/mobile/ios/SyncTrigger/SyncTrigger.swift
//
//  SyncTrigger.swift
//  HomebasePhotos
//
//  Created by Stef Coenen on 19/06/2024.
//

import Foundation
@objc(SyncTrigger)
class SyncTrigger: NSObject {
  @objc
  func runSync() -> Void {
    Task {
      print("[SyncTrigger]: runSync");
      let mediaSync = MediaSync();
      await mediaSync.syncMedia();
    }
  }


  @objc
  func runSingleSync(_ filePath: String,
                     timestampInMs: NSNumber,
                     mimeType: String,
                     identifier: String,
                     width: NSNumber,
                     height: NSNumber,
                     resolver: @escaping RCTPromiseResolveBlock,
                     rejecter: @escaping RCTPromiseRejectBlock) {
    Task {
      print("[SyncTrigger]: runSingleSync");

      do {

        let dotYouClient = try DotYouClient.getDotYouClient()

        let result = try await ImageProvider.uploadMedia(dotYouClient: dotYouClient, filePath: filePath, timestampInMs: Int64(truncating: timestampInMs), mimeType: mimeType, identifier: identifier, width: Int(truncating: width), height: Int(truncating: height), forceLowerQuality: false)

        if(result is SuccessfulUploadResult) {
          // Resolve promise on success
          resolver("Upload completed successfully")
        } else if result is BadRequestUploadResult, let badResult = result as? BadRequestUploadResult, badResult.errorCode == "existingFileWithUniqueId" {
          rejecter("SYNC_ERROR", "File already exists", NSError(domain: "", code: 0))
        } else {
          rejecter("SYNC_ERROR", "Something went wrong", NSError(domain: "", code: 0))
        }

      } catch {
        // Reject promise on error
        rejecter("SYNC_ERROR", "Upload failed with error", error)

        print("[SyncTrigger]: runSingleSync failed with error: \(error)")
      }

      print("[SyncTrigger]: finished");
    }
  }

  @objc
  static func runStaticSync(completion: @escaping (Bool) -> Void) {
    Task {
      print("[SyncTrigger]: runSync");
      let mediaSync = MediaSync();
      await mediaSync.syncMedia();
      completion(true)
    }
  }

  // https://stackoverflow.com/questions/41765634/reactnative-swift-component-how-to-set-the-constructor#answer-50294485
  @objc static func requiresMainQueueSetup() -> Bool {
      return false
  }
}
